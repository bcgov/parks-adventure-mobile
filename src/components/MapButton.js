import React from 'react'
import PropTypes from 'prop-types'
import { Linking, Clipboard } from 'react-native'
import { Portal, useTheme } from 'react-native-paper'
import { Button, Icon, Text, Snackbar } from './MapButton.styles'

function MapButton({ location, absolute = true }) {
  const theme = useTheme()
  const [snackbarVisible, setSnackbarVisibility] = React.useState(false)
  function openMaps() {
    Linking.openURL(
      `https://www.google.com/maps/search/?api=1&query=${location.latitude},${location.longitude}`
    )
  }

  function copyLocation() {
    Clipboard.setString(`${location.latitude},${location.longitude}`)
    setSnackbarVisibility(true)
  }

  return (
    <>
      <Button
        accessibilityLabel={'map-link'}
        onPress={openMaps}
        onLongPress={copyLocation}
        elevation={6}
        absolute={absolute}>
        <Icon name={'map-outline'} size={18} />
        <Text allowFontScaling={false}>Maps</Text>
      </Button>

      <Portal>
        <Snackbar
          theme={{ colors: { accent: theme.colors.secondary50 } }}
          visible={snackbarVisible}
          onDismiss={() => setSnackbarVisibility(false)}
          action={{
            label: 'Ok',
            onPress: () => setSnackbarVisibility(false),
          }}>
          Location copied to Clipboard.
        </Snackbar>
      </Portal>
    </>
  )
}

MapButton.propTypes = {
  location: PropTypes.shape({
    latitude: PropTypes.string.isRequired,
    longitude: PropTypes.string.isRequired,
  }),
  absolute: PropTypes.bool,
}

export default MapButton
