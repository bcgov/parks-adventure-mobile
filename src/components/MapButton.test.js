import React from 'react'
import { render, fireEvent } from '@testing-library/react-native'
import renderSnapshot from 'react-test-renderer'
import { Linking } from 'react-native'
import { Provider } from 'react-native-paper'
import MapButton from './MapButton'

const location = { latitude: '50.987578', longitude: '-119.725971' }

test('Maps button opens url with park lat, long', () => {
  const spy = jest.spyOn(Linking, 'openURL')
  const { getByA11yLabel } = render(
    <Provider>
      <MapButton location={location} />
    </Provider>
  )

  const button = getByA11yLabel('map-link')
  expect(button).toBeDefined()

  fireEvent.press(button)
  expect(spy).toHaveBeenCalledWith(
    `https://www.google.com/maps/search/?api=1&query=${location.latitude},${location.longitude}`
  )

  spy.mockReset()
  spy.mockRestore()
})

test('matches MapButton snapshot', () => {
  const tree = renderSnapshot.create(<MapButton location={location} />).toJSON()
  expect(tree).toMatchSnapshot()
})
