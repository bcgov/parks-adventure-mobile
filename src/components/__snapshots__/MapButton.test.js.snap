// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches MapButton snapshot 1`] = `
<View
  absolute={true}
  accessibilityLabel="map-link"
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "white",
      "borderRadius": 8,
      "elevation": 6,
      "height": 64,
      "paddingTop": 15,
      "position": "absolute",
      "right": 16,
      "shadowColor": "#000000",
      "shadowOffset": Object {
        "height": 5,
        "width": 0,
      },
      "shadowOpacity": 0.24,
      "shadowRadius": 6,
      "top": 100,
      "width": 64,
    }
  }
>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "flexGrow": 1,
        "flexShrink": 1,
      }
    }
  >
    <Text
      allowFontScaling={false}
      index={0}
      siblings={
        Array [
          "Styled(Icon)",
          "Styled(withTheme(Text))",
        ]
      }
      style={
        Array [
          Object {
            "color": undefined,
            "fontSize": 18,
          },
          Array [
            Object {
              "alignSelf": "center",
              "opacity": 0.6,
            },
          ],
          Object {
            "fontFamily": "material-community",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
      total={2}
    >
      Ô¶Å
    </Text>
    <Text
      index={1}
      siblings={
        Array [
          "Styled(Icon)",
          "Styled(withTheme(Text))",
        ]
      }
      style={
        Array [
          Object {
            "color": "#000000",
            "fontFamily": "System",
            "fontWeight": "400",
            "textAlign": "left",
          },
          Array [
            Object {
              "opacity": 0.6,
            },
          ],
        ]
      }
      total={2}
    >
      Maps
    </Text>
  </View>
</View>
`;
